Fix/complete CMake installation

Sarcasm opened the pull request:
This PR tries to address #973 and also fix an issue with a header installed to the wrong place.
I tried to compare the install trees generated by autotools and CMake to make sure there was no functional differences.

I see that @harrishancock said he will do the CMake part here https://github.com/capnproto/capnproto/pull/978#pullrequestreview-397491907, but seeing the imminent release, I thought it was worth trying to do it before the release is official, in case the merge window was still open.
My apology if this feels inappropriate.

kentonv commented on c++/configure.ac:
```diff
@@ -1,6 +1,6 @@
 ## Process this file with autoconf to produce configure.
 
-AC_INIT([Capn Proto],[0.9-dev],[capnproto@googlegroups.com],[capnproto-c++])
+AC_INIT([Capn Proto],[0.8.0-rc1],[capnproto@googlegroups.com],[capnproto-c++])
```
Oops, it looks like you based this PR off the release branch, but are targeting the PR at the master branch. The version in the master branch is now 0.9-dev. Once this is merged into master we can potentially cherry-pick it into a release branch, but please don't change the version in master. (This is causing the test failures.)

kentonv commented:
Thanks for working on this @Sarcasm! Unfortunately this won't make it into the 0.8 release which I'm putting out imminently, but we could potentially do a 0.8.1 release once this is ready.

Sarcasm commented:
No problem for the 0.8.0 release, it was worth a shot but I don't actually need these changes.
Looking forward to this release.

kentonv commented:
Ugh it looks like the TLS bindings might not work quite right on Windows -- there's a test failure on MinGW in the TLS tests. (It looks like the MSVC build isn't detecting OpenSSL, otherwise my guess is it would hit the same problem.)

Though it's kind of amazing that it compiled at all. I think this might be the first time kj-tls has been built on (non-Cygwin) Windows...

Still, we'd need to fix this before releasing. Not sure when I'll have time to dig into it, unfortunately...

harrishancock commented on c++/cmake/CapnProtoConfig.cmake.in:
```diff
@@ -49,7 +49,10 @@ if(NOT _IMPORT_PREFIX)
 endif()
 
-
+if (@WITH_OPENSSL@)
```
If CAPNP_LITE is true, then WITH_OPENSSL could still be "AUTO" at this point, I think. Would that make a lite-mode installation default to depend on OpenSSL? (That said, I wonder if there are any lite-mode installations in the wild, even.)

harrishancock commented on c++/cmake/CapnProtoConfig.cmake.in:
```diff
@@ -49,7 +49,10 @@ if(NOT _IMPORT_PREFIX)
-
+if (@WITH_OPENSSL@)
+  include(CMakeFindDependencyMacro)
+  find_package(OpenSSL REQUIRED)
```
Based on the previous line, should this be `find_dependency()`?

kentonv commented:
Opened #994 where I'm attempting to fix the test failure.

Meanwhile, another thing I noticed is that the MSVC build is failing to find both OpenSSL and Zlib, so does not build against either of them. It'd be really great to fully test the MSVC build. Does anyone have a clue how to install these dependencies for MSVC's use?

harrishancock commented:
GitHub Actions says it has an installation of OpenSSL in Program Files (https://github.com/actions/virtual-environments/blob/master/images/win/Windows2019-Readme.md#openssl)... which apparently CMake specifically searches for: https://github.com/Kitware/CMake/blob/master/Modules/FindOpenSSL.cmake#L108

You could try adding the rest of those directories to the PATH and seeing if one of them gets picked up.

Separately, CMake appears to have a preference for SLProWeb's build of OpenSSL: https://github.com/Kitware/CMake/blob/master/Modules/FindOpenSSL.cmake#L99

http://slproweb.com/products/Win32OpenSSL.html

Could install it and see if it Just Works?

Sarcasm commented:
I'm not familiar with Github Actions (or Windows...), but it looks like OpenSSL and zlib might be available on vcpkg, which is preinstalled on the Windows images IIUC:

- https://github.com/microsoft/vcpkg
- https://github.com/actions/virtual-environments/blob/77ee8ae0f5888316a76ebb829ce014c8e6dcee7e/images/win/Windows2016-Readme.md#vcpkg

SpaceIm commented on c++/src/kj/CMakeLists.txt:
```diff
@@ -163,6 +163,29 @@ if(NOT CAPNP_LITE)
+  target_link_libraries(kj-tls PUBLIC kj-async kj)
+  if (WITH_OPENSSL)
+    target_compile_definitions(kj-tls PRIVATE KJ_HAS_OPENSSL)
+    target_link_libraries(kj-tls PUBLIC OpenSSL::SSL OpenSSL::Crypto)
```
```suggestion
    target_link_libraries(kj-tls PRIVATE OpenSSL::SSL OpenSSL::Crypto)
```
`OpenSSL` headers are not included in `kj-tls` public headers

SpaceIm commented on c++/src/kj/CMakeLists.txt:
```diff
@@ -163,6 +163,29 @@ if(NOT CAPNP_LITE)
+if(NOT CAPNP_LITE)
+  add_library(kj-tls ${kj-tls_sources})
+  add_library(CapnProto::kj-tls ALIAS kj-tls)
+  target_link_libraries(kj-tls PUBLIC kj-async kj)
```
```suggestion
  target_link_libraries(kj-tls PUBLIC kj-async PRIVATE kj)
```

SpaceIm replied:
@Sarcasm why did you remove `kj`?
Even though `kj` is public in `kj-async`, `kj` is still a direct dependency of `kj-tls`, therefore it should be explicitly added in `target_link_libraries` (a target shouldn't rely on public dependencies of its dependencies, if there are also direct dependencies).

Sarcasm replied:
> a target shouldn't rely on public dependencies of its dependencies

Why shouldn't it?

When the code is under control, here a kj lib depending on another kj lib, I don't see how that is a problem.
For me kj-async is kj + async, kj is implied.

Will CMake generate functionally different compile or link commands?

I'm okay to change, but ` target_link_libraries(kj-tls PUBLIC kj-async PRIVATE kj)` feels like being over specifying to me.

Thanks for the review by the way!

SpaceIm replied:
You're right, it doesn't change generated compiler and linker commands, but il will break if `kj` is removed from INTERFACE property of `kj-async`, unless `kj` is added as a dependency of `kj-async`. `kj-tls` should'nt care of `kj-async` properties.
Moreover, removing `kj` makes dependencies between targets harder to understand (because if you do this for this target, maybe others targets follow this practice, therefore maintenairs will be suspicous about CMakeLists).

Sarcasm replied:
We just have to know what is part of the public API.
kj-async declaring kj as a PUBLIC or INTERFACE dependency is a contract with kj users.
Changing that would be an API break.
And if that changes in the headers, it can also be changed in the CMakeLists.txt.

To make a parallel with the STL.
When I use `std::map`, I don't include `utility` for `std::pair` manually, because I assume it's part of the public API of `std::map`.
There is nothing suspicious about that, we just have to find a middle ground.

There is no point in discussing this too much, it might be best to the kj maintainers decide what they want for the project.

harrishancock replied:
In the past, I've operated under the assumption that `kj-*` CMake targets will always publicly depend on `kj`, and `capnp-*` targets will always publicly depend on `capnp`. For example, in samples/CMakeLists.txt, we just link with `CapnProto::capnp-rpc`. So my personal style would have been not to list the explicit `kj` dependency.

That does go against an "include what you use" philosophy, though, which I also like. I would say for the purposes of merging, either one is just fine.

kentonv commented:
https://github.com/capnproto/capnproto/pull/994/commits/629f69115ac19c51fcb7daf2534bc35e5ac0a888 seems to fix tls-test on Win32 (MinGW), if you want to cherry-pick that onto this PR.

So does anyone want to try coaxing the MSVC build into finding OpenSSL (and zlib, while you're at it)? I'm hesitant to try it myself since I'm pretty clueless about both cmake and MSVC build conventions.

Sarcasm commented:
Commit cherry picked.
I will try to add the OpenSSL/zlib support to the Windows build.

kentonv commented:
That's weird, it's still showing the MinGW test failure that my commit was supposed to fix (and did fix in my PR #994...).

kentonv commented:
It looks like the tests actually ran at 14d1ec2 and not at bfd4e29, but bfd4e29 is the tip of the PR... what the heck?

Sarcasm commented:
Yeah, I don't get it.
I pushed your commit first because I wasn't sure if the Github Actions would process all commits or just the last one.
But somehow, it processed something else.

I repushed something, but added a test for OpenSSL/zlib on Windows.

kentonv commented:
Isn't debugging CI fun????

Sarcasm commented:
I love it! Sorry for the noise.
There is one thing I don't understand, it's the MinGW failure, it's still here.

kentonv commented:
@Sarcasm Whoa... I don't know how this happened, but your cherry-picked version of the patch does not match the original. Somehow, the extra code got added to the wrong test case -- my version modifies `KJ_TEST("TLS full duplex")` but your version somehow modifies `KJ_TEST("TLS zero-sized write")`.

kentonv commented:
Oh I see how it happened... my version of the commit was based on your original version of this PR, which was based on master before #989 was merged. It looks like you later rebased your PR to current master, and then applied the cherry-pick, and `patch` managed to get confused and patch the wrong test case because the test cases look very similar and had moved around.

Sarcasm commented:
Wow, thanks for the investigation, I will fix it.

Sarcasm commented:
I think this is working now.
tls and gzip tests were run and passed on Windows.
I switched to Ninja (CMake + Ninja is a nice combo that works on all 3 majors platforms), build logs are now readable, and using this generator nicely fixed some x86 / x64 link issues I got with zlib/openssl.
I used conda to install the dependencies, it comes preinstalled with the virtual images and is faster than vcpkg (which built from source).
I fixed a build issue due to `ssize_t` not available on MSVC, I used `auto`, but maybe `int` was ok too.

kentonv commented:
Hmm what were the problems exactly with the MSVC project build that Ninja fixes?

I worry slightly that some people might depend on the MSVC project files, implying we should test them. On the other hand, we could say if those people really want it, it can be their job to make it work...

Sarcasm commented:
The linking issue was `LNK2019`, e.g.:

    error LNK2019: unresolved external symbol _BIO_set_flags

It complained about zlib and OpenSSL symbols.

I think it was caused by the CMake Visual Studio Generators which defaults to Win32, not x64:

> The default target platform name (architecture) is `Win32`.
> -- https://cmake.org/cmake/help/v3.17/generator/Visual%20Studio%2015%202017.html?highlight=win32#platform-selection

My assumption is that conda zlib and OpenSSL are compiled for x64, but the default generator configuration used Win32. Using `cmake -A x64 ...` may have worked too.

Ninja on the other hand seems to default to the configured MSVC architecture, which I assume is x64.
It's a lot of assumptions to be honest, I wasn't interested in knowing much more about this, the code compiled with cl.exe and linked and that's was sufficient to me.

kentonv commented:
Fair. I might try `cmake -A x64` myself just to see if it happens to work, but you've already done plenty. Thanks so much for working on this!

harrishancock commented:
Nice work @Sarcasm!

Re: Ninja vs. VS generators, both are valuable to support (vscode's CMake Tools extension defaults to the Ninja generator on Windows, for instance). The Visual Studio generator is significantly different from most of the other generators, though, so I think it's worth continuing to CI the VS generator.

kentonv commented:
Doh... I just noticed that the Ninja build is actually building with GCC, not MSVC.

```
-- The CXX compiler identification is GNU 8.1.0
-- Check for working CXX compiler: C:/ProgramData/chocolatey/bin/c++.exe
```

Sarcasm commented:
Ah! Indeed...sorry about that.
While trying to debug this I also see that if the CMake commands fails, the jobs is marked as success:

- https://github.com/Sarcasm/capnproto/runs/817612016
